docs/features/incremental-generators.md

```diff
@@ -0,0 +1,478 @@
+`Values` property of the `context` passed as part of the
+`RegisterExecutionPipeline` callback. Example values sources include
+
+- Compilation
+- AdditionalTexts
+- AnalyzerConfigOptions
+- MetadataReferences
+- ParseOptions
+
+Value sources have 'zero-or-more' potential values that can be produced. For
+example, the `Compilation` will always produce a single value, whereas the
+`AdditionalTexts` will produce a variable number of values, depending on how
+many additional texts where passed to the compiler.
+
+An execution pipeline cannot access these values directly. Instead it supplies a
+set of transforms that will be applied to the data as it changes. Transforms are
+applied through a set of extension methods:
+
+```csharp
+public static partial class IncrementalValueSourceExtensions
+{
+    // 1 => 1 transform 
+    public static IncrementalValueSource<U> Transform<T, U>(this IncrementalValueSource<T> source, Func<T, U> func) => ...
+}
+```
+
+These extension methods allow the user to perform a series of transforms,
+conceptually somewhat similar to LINQ, over the values coming from the data
+source:
+
+```csharp
+initContext.RegisterExecutionPipeline(context =>
+{
+    // get the additional text source
+    IncrementalValueSource<AdditionalText> additionalTexts = context.Sources.AdditionalTexts;
+
+    // apply a 1-to-1 transform on each text, which represents extracting the path
+    IncrementalValueSource<string> transformed = additionalTexts.Transform(static text => text.Path);
+
+});
+```
+
+Note that `transformed` is similarly opaque. It represents the outcome of the
+transformation being applied to the data, but cannot be accessed directly.
+
+The transformed steps can be further transformed, and it is also valid to
+perform multiple transformations on the same input node, essentially 'splitting'
+the pipeline into multiple streams of processing.
+
+```csharp
+    // apply a 1-to-1 transform on each text, extracting the path
+    IncrementalValueSource<string> transformed = additionalTexts.Transform(static text => text.Path);
+
+    // split the processing into two streams of derived data
+    IncrementalValueSource<string> prefixTransform = transformed.Transform(static path => "prefix_" + path);
+    IncrementalValueSource<string> postfixTransform = transformed.Transform(static path => path + "_postfixed");
+```
+
+### Batching
+
+In addition to the 1-to-1 transform shown above, there are also extension
+methods for producing and consuming batches of data. For instance a given
+transform may want to produce more than one value for each input, or want to
+view all the data in a single collected view in order to make cross data
+decisions.
+
+``` csharp
+public static partial class IncrementalValueSourceExtensions
+{
+    // 1 => many (or none)
+    public static IncrementalValueSource<U> TransformMany<T, U>(this IncrementalValueSource<T> source, Func<T, IEnumerable<U>> func) => ...
```

Comment:
nit: in the examples, consider using `TInput` (or `TSource`) and `TResult` for type parameters